### DISCUSSION TITLE
Datum for a NFT smart contract

### DISCUSSION THREAD

**Discussion by ofekron33** [bloxbean/cardano-client-lib#86] [Q&A] [answered]
Hey,
There is an NFT smart contract which can be found in this link:
https://github.com/MartifyLabs/martify.contracts/blob/main/market_testnet_final.plutus

I am trying to use this library in order to use it. For example In order to create a transaction that resembeles  a sale of NFT i need to create  a datum that consists of:
1. Seller address
2. token policy id 
3. token asset name
4. price
5. royaltiesAddress
6. royalties percentage

If i understand correctly, PlutusData is the interface that resemebles the Datum. You have implemented multiple versions of that interface so our datum instance can be either a list,map,big int or ByteString. I thought on using the list plutus data but that is a problem since our datum needs to have multiple different datatypes and lists in java can only hold one datatype.
  
So is it possible to create a PlutusData object that contains multiple different datatypes objects and if  it is then how?

And lastly, each class that implements PlutusData has the function ConstrPlutusData:
![image](https://user-images.githubusercontent.com/58865301/157261960-9ed1cdc0-6fe7-480d-af55-11cf4c0eb630.png)

can you please give us a short explaination as to what is alternative. I have seen in your examples that you give it a value of 0 and i am not sure why.

Thank you.

**Comment by satran004** [comment, marked as answer]
Hi @ofekron33,

>> If i understand correctly, PlutusData is the interface that resemebles the Datum.

Yes, PlutusData is the main interface which is used to represent both Datum and Redeemer data .  And as you mentioned, BigIntPlutusData, BytesPlutusData, ListPlutusData, MapPlutusData, ContrPlutusData are different implementations and only supported types in datum or redeemers according to the spec. 

You can check cbor serialization spec here for more details https://github.com/input-output-hk/cardano-ledger/blob/master/eras/alonzo/test-suite/cddl-files/alonzo.cddl#L274

>> I thought on using the list plutus data but that is a problem since our datum needs to have multiple different datatypes and lists in java can only hold one datatype.

You can use ListPlutusData with different types. It needs PlutusData.

Plz check some examples here https://github.com/bloxbean/cardano-client-lib/blob/6dfbb6085bd1ef8a13891b64f5909739670ad315/src/test/java/com/bloxbean/cardano/client/transaction/spec/ListPlutusDataTest.java#L21


>> can you please give us a short explaination as to what is alternative. I have seen in your examples that you give it a value of 0 and i am not sure why.

"alternative" is used in ConstrPlutusData which is an another implementation of PlutusData which has two fields alternative and list of fields. This represents a constructor in validator code.

If you compare that with cardano-cli's datum json format,  alternative = constructor (constructor's no). Heres' a sample json for cardano-cli compatible datum

```
{"constructor":0,"fields":[{"bytes":"bea1c521df58f4eeef60c647e5ebd88c6039915409f9fd6454a476b9"},{"int":234000000},{"bytes":"74f43bdf645aaeb25f39c6392cdb771ff4eb4da0c017cc183c490b8f"},{"bytes":"63736e66743233"},{"bytes":"bea1c521df58f4eeef60c647e5ebd88c6039915409f9fd6454a476b9"},{"bytes":"00"}]}
```

For above datum, you need to use ConstrPlutusData. 

You can check these tests. How to use CosntrPlutusData 

https://github.com/bloxbean/cardano-client-lib/blob/6dfbb6085bd1ef8a13891b64f5909739670ad315/src/test/java/com/bloxbean/cardano/client/plutus/DefaultPlutusObjectConverterTest.java#L33

You can also use regular POJO with some annotations to define Datum. But make sure to use allowed types for the POJO's fields (Integer, BigInteger, Long, byte[], String)

Check this POJO here https://github.com/bloxbean/cardano-client-lib/blob/6dfbb6085bd1ef8a13891b64f5909739670ad315/src/test/java/com/bloxbean/cardano/client/plutus/DefaultPlutusObjectConverterTest.java#L122

BTW, String type is converted to BytesPlutusData.. so String == byte[]

For your use case (Martify contract), I think you need to use ConstrPlutusData.  If I am not wrong NFTSell is the datum data type here in Haskell.

```
data NFTSale = NFTSale
    { nSeller    :: !PubKeyHash
    , nPrice     :: !Plutus.Integer
    , nCurrency  :: !CurrencySymbol
    , nToken     :: !TokenName
    , nRoyAddr   :: !PubKeyHash
    , nRoyPrct   :: !Plutus.Integer
    } deriving (Generic, ToJSON, FromJSON)
```

So in Java, you need to use ConstrPlutusData with alternative=0 as there is only one constructor. All fields except nPrice and nRoyPrct are bytes.

You can also implement this in POJO with annotation. I will try datum conversion code in Java for this contract and post later to this discussion.

Hope this helps.

**Reply by ofekron33** [reply]
thank you so much this has been really insightful.
